stumpy/stump_tiles.py

```diff
@@ -0,0 +1,794 @@
+    -------
+    None
+
+    Notes
+    -----
+    `DOI: 10.1007/s10115-017-1138-x \
+    <https://www.cs.ucr.edu/~eamonn/ten_quadrillion.pdf>`__
+
+    See Section 4.5
+
+    The above reference outlines a general approach for traversing the distance
+    matrix in a diagonal fashion rather than in a row-wise fashion.
+
+    `DOI: 10.1145/3357223.3362721 \
+    <https://www.cs.ucr.edu/~eamonn/public/GPU_Matrix_profile_VLDB_30DraftOnly.pdf>`__
+
+    See Section 3.1 and Section 3.3
+
+    The above reference outlines the use of the Pearson correlation via Welford's
+    centered sum-of-products along each diagonal of the distance matrix in place of the
+    sliding window dot product found in the original STOMP method.
+    """
+    m_inverse = 1.0 / m
+    constant = (m - 1) * m_inverse * m_inverse  # (m - 1)/(m * m)
+    uint64_m = np.uint64(m)
+
+    iter_ranges = np.zeros((config.STUMPY_N_DIAGONALS, 3), dtype=np.int64)
+    covariances = np.zeros(config.STUMPY_N_DIAGONALS, dtype=np.float64)
+
+    for tile_idx in range(tiles_ranges[thread_idx, 0], tiles_ranges[thread_idx, 1]):
+        (
+            i_offset,  # i location of current tile in distance profile
+            j_offset,  # j location of current tile in distance profile
+            tile_height,  # height of current tile
+            tile_width,  # width of current tile
+            tile_lower_diag,  # index of first diagonal to traverse on current tile
+            tile_upper_diag,  # index of last diagonal to traverse on current tile + 1
+            _,
+        ) = tiles[tile_idx]
+
+        # total number of diagonals to traverse
+        n_diags = tile_upper_diag - tile_lower_diag
+        # number of diagonal "chunks" if we divide our diagonals into chunks
+        n_chunks = int(np.ceil(n_diags / config.STUMPY_N_DIAGONALS))
+
+        # chunk_idx is the index of the chunk we are dealing with
+        for chunk_idx in range(0, n_chunks):
+            # number of diagonals to traverse together
+            chunk_size = min(
+                config.STUMPY_N_DIAGONALS,
+                n_diags - (chunk_idx * config.STUMPY_N_DIAGONALS),
+            )
+
+            # the longest diagonal to be traversed within the current chunk of diagonals
+            longest_diag_len = 0
+            # chunk_diag_idx is the diagonal index we are dealing with,
+            # with respect to the chunk
+            for chunk_diag_idx in range(chunk_size):
+                # tile_diag_idx is the diagonal index we are dealing with,
+                # with respect to the entire tile
+                tile_diag_idx = (
+                    tile_lower_diag
+                    + (chunk_idx * config.STUMPY_N_DIAGONALS)
+                    + chunk_diag_idx
+                )
+
+                # here we store the diagonal index we are dealing with,
+                # with respect to the entire distance matrix
+                iter_ranges[chunk_diag_idx, 0] = tile_diag_idx - i_offset + j_offset
```

Comment:
Again, I wanna avoid intermediate variables. I feel like it would be clearer to me if I just see `- tile_i + tile_j` instead.

Reply:
@alvii147 
I can see your point. One last question before we let this go: 

Do you think calculating `- tile_i + tile_j`  in the first for-loop speeds up the computation?

Reply:
> Do you think calculating `- tile_i + tile_j` in the first for-loop speeds up the computation?

Hmm I would imagine it would save only a minuscule amount of time, but I've seen crazier cases of time optimization while working on this PR, so I'll give this a shot and see if the speed up is significant.